File: Arrays/3.Leaders/Leaders.java
import java.util.*;

class Leaders {
    public static void main(String[] args) {
        int[] arr = {16, 17, 4, 3, 5, 2};
        ArrayList<Integer> result = new ArrayList<>();
        int n = arr.length;

        // Start with the rightmost element
        int maxRight = arr[n - 1];

        // Rightmost element is always a leader
        result.add(maxRight);

        // Traverse the array from right to left
        for (int i = n - 2; i >= 0; i--) {
            if (arr[i] >= maxRight) {
                maxRight = arr[i];
                result.add(maxRight);
            }
        }

        // Reverse the result list to maintain
        // original order
        Collections.reverse(result);

        System.out.println(result);
    }
}

File: 1.Arrays/3.Leaders/Leaders.py
arr = [16, 17, 4, 3, 5, 2]
n = len(arr)

max_from_right = arr[n-1]
leaders = [max_from_right]  # store leaders in a list

for i in range(n-2, -1, -1):
    if arr[i] > max_from_right:
        max_from_right = arr[i]
        leaders.append(arr[i])

# reverse to get original order
leaders.reverse()
print(leaders)
